rust/lance-encoding/src/encodings/physical/block_compress.rs

```diff
@@ -0,0 +1,142 @@
+// SPDX-License-Identifier: Apache-2.0
+// SPDX-FileCopyrightText: Copyright The Lance Authors
+
+use arrow_schema::DataType;
+use snafu::{location, Location};
+use std::{
+    io::{Cursor, Write},
+    str::FromStr,
+};
+
+use lance_core::{Error, Result};
+
+use crate::{
+    data::{DataBlock, OpaqueBlock},
+    encoder::{ArrayEncoder, EncodedArray},
+    format::ProtobufUtils,
+};
+
+pub const COMPRESSION_META_KEY: &str = "lance:compression";
+
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub enum CompressionScheme {
+    None,
+    Zstd,
+}
+
+impl std::fmt::Display for CompressionScheme {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        let scheme_str = match self {
+            Self::Zstd => "zstd",
+            Self::None => "none",
+        };
+        write!(f, "{}", scheme_str)
+    }
+}
+
+impl FromStr for CompressionScheme {
+    type Err = Error;
+
+    fn from_str(s: &str) -> Result<Self> {
+        match s {
+            "none" => Ok(Self::None),
+            "zstd" => Ok(Self::Zstd),
+            _ => Err(Error::invalid_input(
+                format!("Unknown compression scheme: {}", s),
+                location!(),
+            )),
+        }
+    }
+}
+
+pub trait BufferCompressor: std::fmt::Debug + Send + Sync {
+    fn compress(&self, input_buf: &[u8], output_buf: &mut Vec<u8>) -> Result<()>;
+    fn decompress(&self, input_buf: &[u8], output_buf: &mut Vec<u8>) -> Result<()>;
+}
+
+#[derive(Debug, Default)]
+pub struct ZstdBufferCompressor {}
+
+impl BufferCompressor for ZstdBufferCompressor {
+    fn compress(&self, input_buf: &[u8], output_buf: &mut Vec<u8>) -> Result<()> {
+        let mut encoder = zstd::Encoder::new(output_buf, 0)?;
+        encoder.write_all(input_buf)?;
+        match encoder.finish() {
+            Ok(_) => Ok(()),
+            Err(e) => Err(e.into()),
+        }
+    }
+
+    fn decompress(&self, input_buf: &[u8], output_buf: &mut Vec<u8>) -> Result<()> {
+        let source = Cursor::new(input_buf);
+        zstd::stream::copy_decode(source, output_buf)?;
+        Ok(())
+    }
+}
+
+pub struct GeneralBufferCompressor {}
+
+impl GeneralBufferCompressor {
+    pub fn get_compressor(compression_type: &str) -> Box<dyn BufferCompressor> {
+        match compression_type {
+            "" => Box::<ZstdBufferCompressor>::default(),
+            "zstd" => Box::<ZstdBufferCompressor>::default(),
+            _ => panic!("Unsupported compression type: {}", compression_type),
+        }
+    }
+}
+
+// An encoder which uses lightweight compression, such as zstd/lz4 to encode buffers
```

Comment:
`An encoder which uses lightweight compression` might be a mistake? zstd/lz4 are quite computationally heavy